jetty-util/src/main/java/org/eclipse/jetty/util/BlockingCallback.java

```diff
@@ -0,0 +1,82 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2021 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.util;
+
+import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Provides a {@link Callback} that can block the thread
+ * while waiting to be completed.  The block occurs in the
+ * auto-closing of the callback.
+ * <p>
+ * A typical usage pattern is:
+ * <pre>
+ * void someBlockingCall(Object... args) throws IOException
+ * {
+ *     try(BlockingCallback blocker = new BlockingCallback())
+ *     {
+ *         someAsyncCall(args, blocker);
+ *     }
+ * }
+ * </pre>
+ * @see SharedBlockingCallback
+ */
+public class BlockingCallback implements AutoCloseable, Callback, Runnable
```

Comment:
I don't like the `Runnable` implementation as this is mixing concerns that could be handled elsewhere.

Reply:
We have some async methods that just take a runnable, whilst others take a Callback.   Both need this kind of blocking semantic.
I guess we could have a `BlockingRunnable` that only wraps a countdown latch....   Is that the right name?  For now I'll add a TODO to factor out.

Reply:
Note that I've done the same Runnable thing with `SharableBlockingCallback`.    
I think we probably need to refactor all of these classes to be a single Blocker class with inner variations, rather than 3 separate classes.   I've added a TODO.

Reply:
See #7270